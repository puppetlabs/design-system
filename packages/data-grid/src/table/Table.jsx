import React, { Component } from 'react';
import {
  arrayOf,
  shape,
  func,
  any,
  string,
  node,
  bool,
  oneOfType,
} from 'prop-types';
import classnames from 'classnames';
import { get } from 'lodash';
import { Heading, Checkbox, Text, Loading } from '@puppet/react-components';

import ColumnHeader from './ColumnHeader';
import TableHeader from '../tableHeader/TableHeader';
import TableFooter from '../tableFooter/TableFooter';

import './Table.scss';

const propTypes = {
  /** Table data. Must be an array of objects */
  data: arrayOf(shape({})),
  /** Array of column specifications */
  columns: arrayOf(
    shape({
      /** Optional cell data getter method. By default it will grab data at the provided dataKey */
      cellDataGetter: func,
      /** Optional cell renderer method. */
      cellRenderer: func,
      /** Arbitrary additional data passed to the cell renderer for this column */
      // eslint-disable-next-line react/forbid-prop-types
      columnData: any,
      /** Unique string key defining this column */
      dataKey: string.isRequired,
      /** Label for column header text */
      label: node,
      /** Optional feature to make column sortrable */
      sortable: bool,
      /** Optional classname that can be a string or a function taking the dataKey and column index which can be used to render styling on specific column */
      className: oneOfType([func, string]),
      /** Styling for column header text */
      style: shape({}),
    }),
  ).isRequired,
  /** Provides a unique key for each table row. */
  rowKey: oneOfType([func, string]),
  /** Optional function which can be used to render styling on specific rows */
  rowClassName: oneOfType([func, string]),
  /** Optional function which can be used to execute a function on row click, will return rowKey, rowIndex, rowData */
  onRowClick: func,
  /** Render table in fixed-layout mode */
  fixed: bool,
  /** Optional additional table className */
  className: string,
  /** Optional object to decribe the current sorting state for styling */
  sortedColumn: shape({
    /** Descibes sort direction in either asc or desc */
    direction: string,
    /** Descibes the column being sorted using the column dataKey  */
    sortDataKey: string,
  }),
  /** Boolean to determine whether to display loading state */
  loading: bool,
  /** Optional string to provide alternative message when loading */
  loadingMessage: string,
  /** Callback function that will return direction and dataKey on every sort action  */
  onSort: func,
  /** Optional boolean to cause horizontal scrolling when table extends past the container */
  horizontalScroll: bool,
  /** Optional boolean to cause the first column to be fixed when horizontalScroll is true */
  fixedColumn: bool,
  /** Optional boolean to cause the last column to be fixed when horizontalScroll is true */
  fixedLastColumn: bool,
  /** Optional string to provider header which is visable when no data is available */
  emptyStateHeader: string,
  /** Optional string to provider descriptive message explaining the empty state of the table */
  emptyStateMessage: string,
  /** Boolean to render select checkbox column */
  selectable: bool,
  /** Row checked action method, will get checked state and row data  */
  onRowChecked: func,
  /** Action between to the table header checkbox */
  onHeaderChecked: func,
  /** State of the table header checkbox */
  headerCheckState: bool,
  /** Variable passed to the header to say if a bash icon should be used */
  headerIndeterminateState: bool,
};

const defaultProps = {
  data: [],
  rowKey: undefined,
  fixed: false,
  className: '',
  onSort: () => {},
  sortedColumn: { direction: '', sortDataKey: '' },
  horizontalScroll: false,
  fixedColumn: false,
  fixedLastColumn: false,
  emptyStateHeader: 'No data available',
  emptyStateMessage: 'Prompt to action or solution',
  loading: false,
  loadingMessage: 'Loading',
  rowClassName: () => {},
  selectable: false,
  onRowChecked: () => {},
  onHeaderChecked: () => {},
  onRowClick: () => {},
  headerCheckState: false,
  headerIndeterminateState: false,
};

const defaultColumnDefs = {
  cellDataGetter: ({ dataKey, rowData }) => get(rowData, dataKey),
  cellRenderer: ({ cellData }) => cellData,
  label: '',
};

class Table extends Component {
  uniqueIDCheck = (rowKey, rowData, rowIndex) => {
    if (rowKey === undefined) {
      const newRowData = rowData;
      newRowData.id = rowIndex;
      return newRowData.id;
    }
    if (typeof rowKey === 'string') {
      return rowData[rowKey];
    }
    return rowKey(rowData);
  };

  classNameTypeManage = (classname, data, index) => {
    let name;
    if (typeof classname === 'function') {
      name = classname(data, index);
    } else if (typeof classname === 'string') {
      name = classname;
    }
    return name;
  };

  // Allows the child checkbox to be clicked without calling rowClick
  handleOnClick = (e, rowData, rowKey, rowIndex) => {
    const { onRowClick } = this.props;

    if (!e.target.classList.contains('rc-checkbox')) {
      onRowClick(rowKey, rowIndex, rowData);
    }
  };

  render() {
    const {
      data,
      columns,
      fixed,
      rowKey,
      className,
      sortedColumn,
      loading,
      loadingMessage,
      fixedColumn,
      fixedLastColumn,
      horizontalScroll,
      emptyStateHeader,
      emptyStateMessage,
      rowClassName,
      selectable,
      onHeaderChecked,
      headerCheckState,
      onRowChecked,
      onSort,
      headerIndeterminateState,
      onRowClick,
      ...rest
    } = this.props;

    const onColumnSort = (direction, dataKey) => {
      onSort(direction, dataKey);
    };

    return (
      <div
        className={classnames({
          'dg-table-horizontal-scroll': horizontalScroll,
          'dg-table-fixed-column': fixedColumn,
          'dg-table-fixed-last-column': fixedLastColumn,
        })}
      >
        <table
          className={classnames(
            'rc-table',
            { 'rc-table-fixed': fixed },
            className,
          )}
          {...rest}
        >
          <ColumnHeader
            loading={loading}
            loadingMessage={loadingMessage}
            columns={columns}
            selectable={selectable}
            sortedColumn={sortedColumn}
            onSort={onColumnSort}
            onSelectAll={onHeaderChecked}
            selectAllValue={headerCheckState}
            headerIndeterminateState={headerIndeterminateState}
          />
          <tbody>
            {loading && (
              <tr className="rc-table-cell">
                <th
                  className="dg-table-loading-container"
                  colSpan={columns.length}
                >
                  <div className="dg-table-loading-inner-container">
                    <div>
                      <Loading className="dg-loading-size" />
                    </div>
                    <Heading
                      as="h5"
                      color="medium"
                      className="dg-table-loading-header"
                    >
                      {loadingMessage}
                    </Heading>
                  </div>
                </th>
              </tr>
            )}
            {data.map((rowData, rowIndex) => {
              return (
                <tr
                  className={classnames(
                    'dg-table-row',
                    this.classNameTypeManage(rowClassName, rowData, rowIndex),
                    {
                      'dg-table-row-selected':
                        selectable && rowData.selected === true,
                      'dg-table-row-disabled': selectable && !!rowData.disabled,
                    },
                  )}
                  key={this.uniqueIDCheck(rowKey, rowData, rowIndex)}
                  onClick={e =>
                    this.handleOnClick(e, rowData, rowKey, rowIndex)
                  }
                >
                  {selectable ? (
                    <td
                      key={`checkbox ${this.uniqueIDCheck(
                        rowKey,
                        rowData,
                        rowIndex,
                      )}`}
                      className="rc-table-cell"
                    >
                      <Checkbox
                        className={classnames('dg-table-checkbox', {
                          'dg-table-checkbox-disabled':
                            rowData.disabled || 'selectable' in rowData
                              ? !rowData.selectable
                              : false,
                        })}
                        onChange={checked => onRowChecked(checked, rowData)}
                        value={rowData.selected}
                        label=""
                        name=""
                        disabled={
                          rowData.disabled || 'selectable' in rowData
                            ? !rowData.selectable
                            : false
                        }
                      />
                    </td>
                  ) : null}
                  {columns.map((column, columnIndex) => {
                    const {
                      cellDataGetter,
                      cellRenderer,
                      columnData,
                      dataKey,
                      className: columnClassName,
                      style,
                    } = {
                      ...defaultColumnDefs,
                      ...column,
                    };
                    return (
                      <td
                        key={`${(rowIndex, dataKey)}`}
                        className={classnames(
                          'rc-table-cell',
                          this.classNameTypeManage(
                            columnClassName,
                            dataKey,
                            columnIndex,
                          ),
                        )}
                        style={style}
                      >
                        {cellRenderer({
                          cellData: cellDataGetter({
                            dataKey,
                            columnData,
                            rowData,
                          }),
                          columnData,
                          columnIndex,
                          dataKey,
                          rowData,
                          rowIndex,
                        })}
                      </td>
                    );
                  })}
                </tr>
              );
            })}
          </tbody>
        </table>
        {data.length < 1 ? (
          <div className="dg-empty-state-container">
            <Heading className="dg-empty-state-header" as="h3" color="medium">
              {emptyStateHeader}
            </Heading>
            <Text className="dg-empty-state-message" color="medium">
              {emptyStateMessage}
            </Text>
          </div>
        ) : null}
      </div>
    );
  }
}

Table.propTypes = propTypes;
Table.defaultProps = defaultProps;

Table.TableHeader = TableHeader;
Table.TableFooter = TableFooter;

export default Table;
