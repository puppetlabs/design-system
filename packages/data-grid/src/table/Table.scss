@import '@puppet/sass-variables/_index';

.dg-empty-state-container {
  margin: auto;
  text-align: center;
  padding-top: 100px;
  padding-bottom: 100px;

  .dg-empty-state-message {
    margin-top: 8px;
  }
}

/** Column Sorting **/

.dg-column-header-sortable {
  cursor: pointer;
}

.dg-column-header-sortable:hover {
  color: $puppet-b400;
}

.dg-column-header-sortable:focus {
  outline: 0;
  border: 0;
  box-shadow: $puppet-common-focus-outline-inset;

  &::before {
    width: 0;
  }
}

.dg-column-header-sortable:active {
  border: 0;
  box-shadow: $puppet-common-focus-outline-inset;
  color: $puppet-b600;

  &::before {
    width: 0;
  }
}

.dg-column-header-label-active {
  color: $puppet-b500;
}

.dg-column-header-icon-container {
  float: right;
}

.dg-column-header-icon-color {
  &.asc {
    .rc-icon-increment-up {
      fill: $puppet-b500;
    }
  }

  &.desc {
    .rc-icon-increment-down {
      fill: $puppet-b500;
    }
  }
}

/** Row Selection **/

.dg-table-header-checkbox-container {
  width: 16px;
}

.dg-table-header-checkbox > div,
.dg-table-checkbox > div {
  margin-right: 0;
}

.dg-table-checkbox-disabled {
  .rc-checkbox:disabled {
    background: $puppet-n300;
  }
}

.dg-table-row-disabled {
  td {
    color: $puppet-n700;
  }
}

.dg-table-row-selected {
  background-color: $puppet-b50;

  .rc-table-cell::before {
    width: 0;
  }
}

/** Horizontal Scrolling **/

.dg-table-horizontal-scroll {
  overflow-x: scroll;
}

.dg-table-fixed-column {
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;

  th,
  td {
    &:first-child {
      position: -webkit-sticky;
      position: sticky;
      left: 0;
      z-index: 2;
    }
  }

  td:first-child {
    background-color: $puppet-n200;
  }
}

.dg-table-fixed-last-column {
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;

  th,
  td {
    &:last-child {
      position: -webkit-sticky;
      position: sticky;
      right: 0;
      z-index: 2;
    }
  }

  td:last-child {
    background-color: $puppet-white;
  }
}

/** Loading State **/

.dg-table-loading-container {
  width: 100%;
  padding: 5px;
  position: sticky;
  top: 0;
  z-index: 2;
  background-color: $puppet-white;

  .dg-table-loading-inner-container {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .dg-loading-size {
    height: 30px;
    width: 30px;
  }

  .dg-table-loading-header {
    margin-left: 10px;
  }
}
