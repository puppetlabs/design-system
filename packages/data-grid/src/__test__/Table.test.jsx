import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Table from '../table/Table';

configure({ adapter: new Adapter() });

const data = [
  { id: 0, eventType: 'Virus/Malware', affectedDevices: 20, detections: 634 },
  {
    id: 1,
    eventType: 'Spyware/Grayware',
    affectedDevices: 20,
    detections: 634,
  },
  { id: 2, eventType: 'URL Filtering', affectedDevices: 15, detections: 598 },
  { id: 3, eventType: 'Web Reputation', affectedDevices: 15, detections: 598 },
  { id: 4, eventType: 'Network Virus', affectedDevices: 15, detections: 497 },
  {
    id: 5,
    eventType: 'Application Control',
    affectedDevices: 0,
    detections: 0,
  },
  {
    id: 6,
    eventType: 'Application Control',
    affectedDevices: 0,
    detections: 0,
  },
];

const columns = [
  {
    label: 'Event Type',
    dataKey: 'eventType',
    className: 'testColumnClassName',
  },
  {
    label: 'Affected Devices',
    dataKey: 'affectedDevices',
    className: 'testColumnClassName',
  },
  {
    label: 'Detections',
    dataKey: 'detections',
    className: 'testColumnClassName',
  },
  {
    label: 'Linked field',
    dataKey: 'Link',
    className: 'testColumnClassName',
  },
];

const sortedColumns = [
  {
    label: 'Event Type',
    dataKey: 'eventType',
    className: 'testColumnClassName',
    sortable: true,
  },
  {
    label: 'Affected Devices',
    dataKey: 'affectedDevices',
    className: 'testColumnClassName',
    sortable: true,
  },
  {
    label: 'Detections',
    dataKey: 'detections',
    className: 'testColumnClassName',
    sortable: true,
  },
  {
    label: 'Linked field',
    dataKey: 'Link',
    className: 'testColumnClassName',
    sortable: true,
  },
];

const wrapper = mount(<Table columns={columns} data={data} />);
const wrapper2 = <Table columns={columns} />;
const wrapper3 = mount(<Table columns={columns} data={[]} />);
const wrapper4 = mount(<Table columns={sortedColumns} data={data} />);
const wrapper6 = mount(
  <Table columns={columns} data={data} rowClassName="testRowClassName" />,
);
const wrapper7 = mount(<Table columns={columns} data={data} loading />);

describe('Snapshot test', () => {
  test('Check component matches previous HTML snapshot', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper3).toMatchSnapshot();
  });
});

describe('Data Prop', () => {
  test('Check data prop default value', () => {
    const props = mount(wrapper2).props();
    expect(props.data).toEqual([]);
  });

  test('Empty array should render empty state', () => {
    expect(
      wrapper3
        .find('div.dg-empty-state-container')
        .hasClass('dg-empty-state-container'),
    ).toEqual(true);
  });

  test('that when table is loading it renders loading state', () => {
    expect(
      wrapper7
        .find('div.dg-table-loading-inner-container')
        .hasClass('dg-table-loading-inner-container'),
    ).toEqual(true);
  });

  test('Table should render correct number of rows', () => {
    expect(wrapper.find('tr.dg-table-row')).toHaveLength(7);
  });

  test('Data values should render in correct column', () => {
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">Virus/Malware</td>),
    );
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">20</td>),
    );
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">634</td>),
    );
  });
});

describe('Column Prop', () => {
  test('All columns render', () => {
    expect(wrapper.find('tr.rc-table-header').children()).toHaveLength(4);
  });

  test('Correct labels displayed / Columns render in correct order', () => {
    expect(wrapper.find('tr.rc-table-header').childAt(0).text()).toBe(
      columns[0].label,
    );

    expect(wrapper.find('tr.rc-table-header').childAt(1).text()).toBe(
      columns[1].label,
    );

    expect(wrapper.find('tr.rc-table-header').childAt(2).text()).toBe(
      columns[2].label,
    );

    expect(wrapper.find('tr.rc-table-header').childAt(3).text()).toBe(
      columns[3].label,
    );
  });

  const data2 = [
    {
      eventType: 'Virus/Malware',
      eventResults: { affectedDevices: 20, detections: 600, unique: 40 },
    },
  ];

  const columns2 = [
    {
      label: 'Event Type1',
      dataKey: 'eventType',
    },
    {
      label: 'Affected Devices',
      dataKey: 'eventResults.affectedDevices',
    },
    { label: 'Detections', dataKey: 'eventResults.detections' },
    { label: 'Linked field', dataKey: 'eventResults.unique' },
  ];

  const wrapper8 = mount(<Table data={data2} columns={columns2} />);

  test('Customs can collect data with dataKey paths', () => {
    expect(wrapper8.find('td.rc-table-cell').at(2).text()).toEqual('600');
  });
});

describe('Sortable Props', () => {
  const mockfunc = jest.fn();
  const sortedState = { direction: 'asc', sortDataKey: 'eventType' };
  const wrapper5 = mount(
    <Table
      columns={sortedColumns}
      data={data}
      onSort={mockfunc}
      sortedColumn={sortedState}
    />,
  );

  test('When sortable prop is true header icons render ', () => {
    expect(wrapper4.find('.rc-icon-increment')).toHaveLength(columns.length);
  });

  test('SortedColumn renders correctly', () => {
    expect(
      wrapper5
        .find('.rc-table-header')
        .children()
        .exists('.dg-column-header-label-active'),
    ).toBe(true);
  });
});

describe('Custom classes', () => {
  const rowMockfunc = jest.fn();
  const columnMockfunc = jest.fn();
  // eslint-disable-next-line
  const wrapper7 = mount(
    <Table
      columns={columns.map((x) => ({
        ...x,
        className: () => columnMockfunc(),
      }))}
      data={data}
      rowClassName={rowMockfunc}
    />,
  );

  test('Custom row classname of string is rendered', () => {
    wrapper6.find('tr.dg-table-row').forEach((node) => {
      expect(node.hasClass('testRowClassName')).toEqual(true);
    });
  });

  test('Custom row classname of function is called', () => {
    expect(rowMockfunc).toHaveBeenCalledTimes(7);
  });

  test('Custom column classname of string is rendered', () => {
    wrapper6.find('td.rc-table-cell').forEach((node) => {
      expect(node.hasClass('testColumnClassName')).toEqual(true);
    });
  });

  test('Custom column classname of function is called', () => {
    expect(columnMockfunc).toHaveBeenCalledTimes(28);
  });
});

describe('Selection Props', () => {
  const rowCheckMockfunc = jest.fn();
  const rowClickMockfunc = jest.fn();
  const wrapper9 = mount(
    <Table
      columns={columns}
      data={data}
      selectable
      onRowChecked={rowCheckMockfunc}
      onRowClick={rowClickMockfunc}
    />,
  );

  test('When selectable prop passed table renders correctly', () => {
    expect(
      wrapper9
        .find('.dg-table-header-checkbox')
        .first()
        .parent()
        .is('th.dg-table-header-checkbox-container'),
    ).toEqual(true);
  });
  expect(wrapper9.findWhere((n) => n.name() === 'Checkbox')).toHaveLength(8);

  test('When row checkbox is clicked fire onUpdateData', () => {
    wrapper9.find('.rc-checkbox').last().simulate('change');
    expect(rowCheckMockfunc).toHaveBeenCalled();
  });

  test('When property `selected` is true in data prop object table row cells are rendered correctly', () => {
    const data3 = [
      {
        eventType: 'Virus/Malware',
        eventResults: { affectedDevices: 20, detections: 600, unique: 40 },
        selected: true,
      },
    ];

    const columns3 = [
      {
        label: 'Event Type1',
        dataKey: 'eventType',
      },
      {
        label: 'Affected Devices',
        dataKey: 'eventResults.affectedDevices',
      },
      { label: 'Detections', dataKey: 'eventResults.detections' },
      { label: 'Linked field', dataKey: 'eventResults.unique' },
    ];

    const wrapper11 = mount(
      <Table data={data3} columns={columns3} selectable />,
    );
    expect(wrapper11.find('.rc-checkbox').last().prop('checked')).toEqual(true);
  });

  test('When row is clicked fire onRowClick', () => {
    wrapper9.find('.dg-table-row').first().simulate('click');

    expect(rowClickMockfunc).toHaveBeenCalledWith(undefined, 0, {
      affectedDevices: 20,
      detections: 634,
      eventType: 'Virus/Malware',
      id: 0,
    });
  });
});
