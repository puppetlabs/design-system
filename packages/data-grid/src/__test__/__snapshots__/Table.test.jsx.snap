// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Check component matches previous HTML snapshot 1`] = `
<Table
  className=""
  columns={
    [
      {
        "className": "testColumnClassName",
        "dataKey": "eventType",
        "label": "Event Type",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "affectedDevices",
        "label": "Affected Devices",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "detections",
        "label": "Detections",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "Link",
        "label": "Linked field",
      },
    ]
  }
  data={
    [
      {
        "affectedDevices": 20,
        "detections": 634,
        "eventType": "Virus/Malware",
        "id": 0,
      },
      {
        "affectedDevices": 20,
        "detections": 634,
        "eventType": "Spyware/Grayware",
        "id": 1,
      },
      {
        "affectedDevices": 15,
        "detections": 598,
        "eventType": "URL Filtering",
        "id": 2,
      },
      {
        "affectedDevices": 15,
        "detections": 598,
        "eventType": "Web Reputation",
        "id": 3,
      },
      {
        "affectedDevices": 15,
        "detections": 497,
        "eventType": "Network Virus",
        "id": 4,
      },
      {
        "affectedDevices": 0,
        "detections": 0,
        "eventType": "Application Control",
        "id": 5,
      },
      {
        "affectedDevices": 0,
        "detections": 0,
        "eventType": "Application Control",
        "id": 6,
      },
    ]
  }
  emptyStateHeader="No data available"
  emptyStateMessage="Prompt to action or solution"
  fixed={false}
  fixedColumn={false}
  fixedLastColumn={false}
  headerCheckState={false}
  headerIndeterminateState={false}
  horizontalScroll={false}
  loading={false}
  loadingMessage="Loading"
  onHeaderChecked={[Function]}
  onRowChecked={[Function]}
  onRowClick={[Function]}
  onSort={[Function]}
  rowClassName={[Function]}
  selectable={false}
  sortedColumn={
    {
      "direction": "",
      "sortDataKey": "",
    }
  }
>
  <div
    className=""
  >
    <table
      className="rc-table"
    >
      <ColumnHeader
        columns={
          [
            {
              "className": "testColumnClassName",
              "dataKey": "eventType",
              "label": "Event Type",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "affectedDevices",
              "label": "Affected Devices",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "detections",
              "label": "Detections",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "Link",
              "label": "Linked field",
            },
          ]
        }
        headerIndeterminateState={false}
        loading={false}
        loadingMessage="Loading"
        onSelectAll={[Function]}
        onSort={[Function]}
        selectAllValue={false}
        selectable={false}
        sortedColumn={
          {
            "direction": "",
            "sortDataKey": "",
          }
        }
      >
        <thead>
          <tr
            className="rc-table-header"
          >
            <th
              className="rc-table-header-cell"
              key="eventType"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Event Type
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="affectedDevices"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Affected Devices
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="detections"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Detections
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="Link"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Linked field
              </span>
            </th>
          </tr>
        </thead>
      </ColumnHeader>
      <tbody>
        <tr
          className="dg-table-row"
          key="0"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Virus/Malware
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            20
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            634
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="1"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Spyware/Grayware
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            20
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            634
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="2"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            URL Filtering
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            598
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="3"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Web Reputation
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            598
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="4"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Network Virus
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            497
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="5"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Application Control
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="6"
          onClick={[Function]}
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Application Control
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
      </tbody>
    </table>
  </div>
</Table>
`;

exports[`Snapshot test Check component matches previous HTML snapshot 2`] = `
<Table
  className=""
  columns={
    [
      {
        "className": "testColumnClassName",
        "dataKey": "eventType",
        "label": "Event Type",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "affectedDevices",
        "label": "Affected Devices",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "detections",
        "label": "Detections",
      },
      {
        "className": "testColumnClassName",
        "dataKey": "Link",
        "label": "Linked field",
      },
    ]
  }
  data={[]}
  emptyStateHeader="No data available"
  emptyStateMessage="Prompt to action or solution"
  fixed={false}
  fixedColumn={false}
  fixedLastColumn={false}
  headerCheckState={false}
  headerIndeterminateState={false}
  horizontalScroll={false}
  loading={false}
  loadingMessage="Loading"
  onHeaderChecked={[Function]}
  onRowChecked={[Function]}
  onRowClick={[Function]}
  onSort={[Function]}
  rowClassName={[Function]}
  selectable={false}
  sortedColumn={
    {
      "direction": "",
      "sortDataKey": "",
    }
  }
>
  <div
    className=""
  >
    <table
      className="rc-table"
    >
      <ColumnHeader
        columns={
          [
            {
              "className": "testColumnClassName",
              "dataKey": "eventType",
              "label": "Event Type",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "affectedDevices",
              "label": "Affected Devices",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "detections",
              "label": "Detections",
            },
            {
              "className": "testColumnClassName",
              "dataKey": "Link",
              "label": "Linked field",
            },
          ]
        }
        headerIndeterminateState={false}
        loading={false}
        loadingMessage="Loading"
        onSelectAll={[Function]}
        onSort={[Function]}
        selectAllValue={false}
        selectable={false}
        sortedColumn={
          {
            "direction": "",
            "sortDataKey": "",
          }
        }
      >
        <thead>
          <tr
            className="rc-table-header"
          >
            <th
              className="rc-table-header-cell"
              key="eventType"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Event Type
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="affectedDevices"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Affected Devices
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="detections"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Detections
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="Link"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Linked field
              </span>
            </th>
          </tr>
        </thead>
      </ColumnHeader>
      <tbody />
    </table>
    <div
      className="dg-empty-state-container"
    >
      <Heading
        allCaps={false}
        as="h3"
        className="dg-empty-state-header"
        color="medium"
        hero={false}
        label={false}
      >
        <h3
          className="rc-heading rc-heading-h3 rc-heading-medium dg-empty-state-header"
        >
          No data available
        </h3>
      </Heading>
      <Text
        as="div"
        className="dg-empty-state-message"
        color="medium"
        size="medium"
        style={{}}
      >
        <div
          className="rc-text rc-text-size-medium rc-text-medium dg-empty-state-message"
          style={{}}
        >
          Prompt to action or solution
        </div>
      </Text>
    </div>
  </div>
</Table>
`;
